Javascript Demux has barebones functionality, need to do some refactoring on demo.js to smooth out behavior.

brianxautumn opened the pull request:
Js Demuxer is commented out in OGVWrapperCodec.js, can uncomment if you like for testing.

(Seek is not yet fully implemented, the cues do in fact load but getting the right behavior for it to do this efficiently is tricky right now). 

Check out the bottom of WebmDemuxer.js . Need to do a little bit of refactoring on demo.js but not sure exactly how you would like to do it.

My new idea is to separate the behaviors seek, vs scrub. seek should be a direct jump to a keypoint offset which can be triggered either by switching streams to different resolution, and staying at time offset, or once chapters are implementing , jumping directly to a chapter, jump to beginning, ect.

A scrub event can be while the user scrubs the thumb, the first scrub should trigger load cues in the background which should attempt to load entire cues.

When thumb is left off, scrubEnd which will reinit the stream at the desired offset, if cues is still loading, save the offset in a temp variable then as soon as the next process loop comes around flush and reinit here. Will only need to do this on first scrub. Can also use this event to try to preview the scrub either by brute force rendering, or using a sprite.

Feel free to edit anything you like, still needs a lot of cleaning up.

brion commented:
This is super exciting! :)

A few quick notes on getting this integrated:
* we should use babel (and if necessary webpack) to pack the demuxer into its own module that can be loaded in IE 10/11 as well as newer browsers with native ES6 bits. (In future as I modularize things it'll be much easier to make a custom 'pre-composed' distribution for an application, but for now let's stick with the dynamic web loader)
* once the pure-JS demuxer is at full feature parity, we can probably drop the nestegg one :D
* while testing, there should be an options key on the Player to select which to load, and a selection in the player mode in the demo harness

brianxautumn commented:
> we should use babel (and if necessary webpack) to pack the demuxer into its own module that can be loaded in IE 10/11 as well as newer browsers with native ES6 bits. (In future as I modularize things it'll be much easier to make a custom 'pre-composed' distribution for an application, but for now let's stick with the dynamic web loader)

Cool, i'll get to that as soon as I get the scrub done to A/B it.

> once the pure-JS demuxer is at full feature parity, we can probably drop the nestegg one :D
👍 

> while testing, there should be an options key on the Player to select which to load, and a selection in the player mode in the demo harness
Do you mean for switching back and forth between nestegg/js?


Lot's of new notes:

1. Will be adding option to force preload cues meta/chapters. Why? In case chapters are in fact available but located at the end, you can specify to check for them and add them in before playback.

2. I wanted to add in a bunch of new interfacing from the player to the wrapper in order to access methods on the demuxer, but the way it is now, a lot of the methods and properties are private. Rewrote OGVWrapperCodec to ES6 syntax, made everything public for now, but now interfacing with webmdemuxer is much easier without having to do things like:

`this.demuxer.onseek = function (offset) {
                    if (this.onseek) {
                        this.onseek(offset);
                    }`

3. Callbacks are really deep, I'm considering bluebird promise library. Wanted your input on this.

4. I fixed some merge overwriting so it does play again with the YUV version, have not pushed it to my live demo server, I will when I get this working.  

5. I'll hopefully have a working scrub for you within a few days to try.

6. The live demo with nestegg seek still fails for me sometimes. Have not investigated why, but it's hit or miss right now.

7. Made a tiny demo, haven't uploaded it yet, but I have a bool decoder class, and I can parse some of the vp8 headers!

seek branch:
https://github.com/brianxautumn/ogv.js/tree/Demuxer-Development-Seek
